### PR TITLE
Psp refactor

### PR DESCRIPTION
**Pull request by antoine-levitt**
No functional change (hopefully)

### CODE REVIEW

File: src/pseudo/NormConservingPsp.jl [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import Base.Broadcast.broadcastable
+Base.Broadcast.broadcastable(psp::NormConservingPsp) = Ref(psp)
+
+function projector_indices(psp::NormConservingPsp)
+    ((i, l, m) for l in 0:psp.lmax for i in 1:size(psp.h[l+1], 1)
+     for m = -l:l)
+end
+
+eval_psp_energy_correction(psp::NormConservingPsp, n_electrons) =
+    eval_psp_energy_correction(Float64, psp, n_electrons)
```
Comment (mfherbst) [inline src/pseudo/NormConservingPsp.jl:80]: Move that up to the other instance of this function.

File: src/pseudo/NormConservingPsp.jl [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+is the number of electrons per unit cell. This defines the uniform compensating background
+charge, which is assumed here.
+
+Notice: The returned result is the *energy per unit cell* and not the energy per volume.
+To obtain the latter, the caller needs to divide by the unit cell volume.
+"""
+eval_psp_energy_correction(T, psp::NormConservingPsp, n_electrons) = zero(T)
+# by default, no correction
+
+
```
Comment (mfherbst) [inline src/pseudo/NormConservingPsp.jl:69]: I think 3 empty lines is a bit too many ...

File: src/pseudo/NormConservingPsp.jl [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+eval_psp_energy_correction(psp::NormConservingPsp, n_electrons) =
+    eval_psp_energy_correction(Float64, psp, n_electrons)
+
+# Number of projection vectors per atom
+function count_n_proj(psp::NormConservingPsp)
+    psp.lmax < 0 ? 0 : sum(size(psp.h[l + 1], 1) * (2l + 1) for l in 0:psp.lmax)::Int
+end
+function count_n_proj(atoms)
+    sum(count_n_proj(psp)*length(positions) for (psp, positions) in atoms)::Int
```
Comment (mfherbst) [inline src/pseudo/NormConservingPsp.jl:87]: Space around `*`.

File: src/pseudo/NormConservingPsp.jl [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+# Subtypes must implement the following:
+
+#### Fields:
+
+#    lmax::Int                   # Maximal angular momentum in the non-local part
+#    h::Vector{Matrix{Float64}}  # Projector coupling coefficients per AM channel: h[l][i1,i2]
+#    Zion::Int                   # Ionic charge (Z - valence electrons)
+#    identifier::String          # String identifying the PSP
+#    description::String         # Descriptive string
+
```
Comment (mfherbst) [inline src/pseudo/NormConservingPsp.jl:11]: Refer to PspHgh as an example?

File: src/pseudo/NormConservingPsp.jl [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+#### Methods:
+"""
+    eval_psp_projector_real(psp::PspHgh, i, l, q::Real)
+
+Evaluate the radial part of the `i`-th projector for angular momentum `l`
+in real-space at the vector with modulus `r`.
+"""
+eval_psp_projector_real(psp::NormConservingPsp, i, l, r::T) where {T <: Real} =
+    error("Not implemented")
+eval_psp_projector_real(psp::NormConservingPsp, i, l, r::AbstractVector) = eval_psp_projector_real(psp, i, l, norm(r))
```
Comment (mfherbst) [inline src/pseudo/NormConservingPsp.jl:21]: line break

File: src/pseudo/NormConservingPsp.jl [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+#    description::String         # Descriptive string
+
+#### Methods:
+"""
+    eval_psp_projector_real(psp::PspHgh, i, l, q::Real)
+
+Evaluate the radial part of the `i`-th projector for angular momentum `l`
+in real-space at the vector with modulus `r`.
+"""
+eval_psp_projector_real(psp::NormConservingPsp, i, l, r::T) where {T <: Real} =
```
Comment (mfherbst) [inline src/pseudo/NormConservingPsp.jl:19]: Just `r::Real` is sufficient.

File: src/pseudo/NormConservingPsp.jl [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+#    lmax::Int                   # Maximal angular momentum in the non-local part
+#    h::Vector{Matrix{Float64}}  # Projector coupling coefficients per AM channel: h[l][i1,i2]
+#    Zion::Int                   # Ionic charge (Z - valence electrons)
+#    identifier::String          # String identifying the PSP
+#    description::String         # Descriptive string
+
+#### Methods:
+"""
+    eval_psp_projector_real(psp::PspHgh, i, l, q::Real)
```
Comment (mfherbst) [inline src/pseudo/NormConservingPsp.jl:14]: Replace the `q` by `r` here for consistency.

File: src/pseudo/NormConservingPsp.jl [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    eval_psp_projector_fourier(psp::PspHgh, i, l, q::Number)
+
+Evaluate the radial part of the `i`-th projector for angular momentum `l`
+at the reciprocal vector with modulus `q`:
+p(q) = ∫_R^3 p_{il}(r) e^{-iqr} dr
+     = 4π ∫_{R+} r^2 p_{il}(r) j_l(q r) dr
+"""
+eval_psp_projector_fourier(psp::NormConservingPsp, i, l, q::T) where {T <: Real} =
+    error("Not implemented")
+eval_psp_projector_fourier(psp::NormConservingPsp, q::AbstractVector) = eval_psp_projector_fourier(psp, norm(q))
```
Comment (mfherbst) [inline src/pseudo/NormConservingPsp.jl:33]: Line break.

File: src/pseudo/NormConservingPsp.jl (hunk 6 above)
Comment (mfherbst) [reply src/pseudo/NormConservingPsp.jl:19]: Same at other places in this file.

File: src/pseudo/NormConservingPsp.jl [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+"""
+    eval_psp_local_fourier(psp, q)
+
+Evaluate the local part of the pseudopotential in reciprocal space:
+V(q) = ∫_R^3 Vloc(r) e^{-iqr} dr
+     = 4π ∫_{R+} sin(qr)/q r e^{-iqr} dr
+"""
+eval_psp_local_fourier(psp::NormConservingPsp, q::T) where {T <: Real} = 
+    error("Not implemented")
+eval_psp_local_fourier(psp::NormConservingPsp, q::AbstractVector) = eval_psp_local_fourier(psp, norm(q))
```
Comment (mfherbst) [inline src/pseudo/NormConservingPsp.jl:53]: Line break.

File: src/pseudo/NormConservingPsp.jl (hunk 2 above)
Comment (antoine-levitt) [reply src/pseudo/NormConservingPsp.jl:69]: I figured you would :-) In general you don't like to delimitate visually parts of files (here, the methods to be implemented and the methods you get for free), but I'm not sure why? It does help reading.

File: src/pseudo/NormConservingPsp.jl (hunk 6 above)
Comment (antoine-levitt) [reply src/pseudo/NormConservingPsp.jl:19]: good point

File: src/pseudo/NormConservingPsp.jl [hunk 10]
```diff
@@ -48,9 +53,10 @@ Evaluate the local part of the pseudopotential in reciprocal space:
 V(q) = ∫_R^3 Vloc(r) e^{-iqr} dr
      = 4π ∫_{R+} sin(qr)/q r e^{-iqr} dr
 """
-eval_psp_local_fourier(psp::NormConservingPsp, q::T) where {T <: Real} = 
+eval_psp_local_fourier(psp::NormConservingPsp, q) =
```
Comment (mfherbst) [inline src/pseudo/NormConservingPsp.jl:56]: Not sure it matters, but here you dropped the `::Real`. I'd add it for consistency.

File: src/pseudo/NormConservingPsp.jl (hunk 2 above)
Comment (mfherbst) [reply src/pseudo/NormConservingPsp.jl:76]: I think when it comes to empty lines 1 or 2 is good. If it's more that you should better put a short comment why you separate stuff (like a headline in the file).

Comment (antoine-levitt) [reply src/pseudo/NormConservingPsp.jl:76]: ok

File: src/pseudo/NormConservingPsp.jl (hunk 10 above)
Comment (antoine-levitt) [reply src/pseudo/NormConservingPsp.jl]: woops you're right, thanks

**Pull request merged**